src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala

```diff
@@ -111,6 +113,77 @@ trait EntityComponent {
+            sql"""left outer join ENTITY_ATTRIBUTE sort_a on sort_a.owner_id = e.id and sort_a.name = $sortFieldName and ifnull(sort_a.list_index, 0) = 0 left outer join ENTITY sort_e_ref on sort_a.value_entity_ref = sort_e_ref.id """)
+        }
+
+        concatSqlActions(sql"""select e.id, e.name, e.all_attribute_values #$sortColumns from ENTITY e """, sortJoin, sql""" where e.entity_type = $entityType and e.workspace_id = $workspaceId """)
+      }
+
+      def actionForPagination(workspaceContext: SlickWorkspaceContext, entityType: String, entityQuery: model.EntityQuery) = {
+        /*
+        The query here starts with baseEntityAndAttributeSql which is the typical select
+        to pull entities will all attributes and references. A join is added on a sub select from ENTITY
+        (and ENTITY_ATTRIBUTE if there is a sort field other than name) which constrains to only entities of the
+        right type, workspace, and matches the filters then sorts and slices out the appropriate page of entities.
+         */
+
+        def filterSql(prefix: String, alias: String) = {
+          val filtersOption = entityQuery.filterTerms.map { _.split(" ").toSeq.map { term =>
+            sql"concat(#$alias.name, ' ', #$alias.all_attribute_values) like ${'%' + term.toLowerCase + '%'}"
+          }}
+
+          filtersOption match {
+            case None => sql""
+            case Some(filters) =>
+              concatSqlActions(sql"#$prefix ", reduceSqlActionsWithDelim(filters, sql" and "))
+          }
+        }
+
+        def order(alias: String) = entityQuery.sortField match {
+          case "name" => sql" order by #$alias.name #${SortDirections.toSql(entityQuery.sortDirection)} "
+          case _ => sql" order by #$alias.sort_list_length #${SortDirections.toSql(entityQuery.sortDirection)}, #$alias.sort_field_string #${SortDirections.toSql(entityQuery.sortDirection)}, #$alias.sort_field_number #${SortDirections.toSql(entityQuery.sortDirection)}, #$alias.sort_field_boolean #${SortDirections.toSql(entityQuery.sortDirection)}, #$alias.sort_field_ref #${SortDirections.toSql(entityQuery.sortDirection)}, #$alias.name #${SortDirections.toSql(entityQuery.sortDirection)} "
+        }
+
+        val paginationJoin = concatSqlActions(
+          sql""" join (select * from (""",
+          paginationSubquery(workspaceContext.workspaceId, entityType, entityQuery.sortField),
+          sql") pagination ",
+          filterSql("where", "pagination"),
+          order("pagination"),
+          sql" limit #${entityQuery.pageSize} offset #${(entityQuery.page-1) * entityQuery.pageSize} ) p on p.id = e.id "
+        )
+
+        for {
+          filteredCount <- concatSqlActions(sql"select count(1) from (", paginationSubquery(workspaceContext.workspaceId, entityType, entityQuery.sortField), sql") pagination ", filterSql("where", "pagination")).as[Int]
```

Comment:
if you think it is minor I would prefer to leave it as is. Perhaps a better optimization is to return the filtered and unfiltered counts in separate api calls then only call them as needed. There is no need to get the counts all the time. The unfiltered count changes only when the data changes and the filtered count changes only when the filter changes. But that would be an api change that I am not willing to do yet.